Env var sources / downward API

pmorie opened the pull request:
As discussed in #6000 and #2316, relevant also to #386.

pmorie commented:
@bgrant0607 @smarterclayton @erictune @thockin @derekwaynecarr @mfojtik PTAL.

Preliminary WIP for comment.  I don't think there's any controversy about the concepts here, based on the last community hangout.  Looking for feedback on the factorings.

smarterclayton commented on pkg/kubelet/kubelet.go:
```diff
@@ -771,6 +777,43 @@ func (kl *Kubelet) makeEnvironmentVariables(ns string, container *api.Container)
+		switch {
+		case envVar.ObjectInfo != nil:
+			objInfo := envVar.ObjectInfo
+			internalFieldPath, _, err := api.Scheme.ConvertFieldLabel(objInfo.APIVersion, "Pod", objInfo.FieldPath, "")
```
add the `TODO: use Kind from ObjectInfo` so it doesn't confuse people looking at this.

pmorie commented:
Taking this a little further today in terms of factoring within the kubelet.  If anyone would like info added to the skeletal API, speak out.

I'll making the versioned API changes once people have a chance to chime in here.

pmorie commented:
@sdminonne fyi

sdminonne commented:
@pmorie thanks

derekwaynecarr commented:
@pmorie - my first impression is that I am good with this approach.  my initial reaction was unsure if i liked the metadata.* syntax, but I have no other good alternative in mind so that lgtm.

ironcladlou commented:
Is there any reason to place security restrictions on the host source? Is it okay that any user can access host information via this mechanism? Right now it's limited to hostname, but I assume the set of fields will increase over time.

pmorie commented on pkg/api/testing/fuzzer.go:
```diff
@@ -191,6 +191,21 @@ func FuzzerFor(t *testing.T, version string, src rand.Source) *fuzz.Fuzzer {
 		},
+		func(ev *api.EnvVar, c fuzz.Continue) {
+			ev.Name = c.RandString()
+			if c.RandBool() {
```
I have an idea to rework this and make it a bit better.

pmorie commented on pkg/api/serialization_test.go:
```diff
@@ -164,6 +164,20 @@ func TestEncode_Ptr(t *testing.T) {
 		Spec: api.PodSpec{
 			RestartPolicy: api.RestartPolicyAlways,
 			DNSPolicy:     api.DNSClusterFirst,
+			Containers: []api.Container{
```
Debugging cruft; will remove

pmorie commented:
@bgrant0607 This is shaping up; what supplemental stuff does it need to go in with? I can think of:

1.  Pass through the examples to see where the downward API can be used instead
2.  Docs
3.  E2E test

@smarterclayton @thockin @derekwaynecarr any opinions on the above?

bgrant0607 commented on pkg/api/v1beta3/types.go:
```diff
@@ -509,6 +509,33 @@ type EnvVar struct {
 	// Optional: defaults to "".
 	Value string `json:"value,omitempty" description:"value of the environment variable; defaults to empty string"`
+	// Optional: specify a source for the value of the EnvVar.
+	*EnvVarSource `json:"source,omitempty" description:"source for the environment variable's value; cannot be used if value is set"`
```
A pointer to an embedded struct is confusing. It needs to not be embedded, which would make sense because the json isn't embedded.

bgrant0607 replied:
Re. the description: We can't tell whether value is "set", only whether it is empty.

pmorie replied:
@bgrant0607 Are you suggesting instead:

```go
// EnvVar represents an environment variable present in a Container.
type EnvVar struct {
	// Required: This must be a C_IDENTIFIER.
	Name string `json:"name" description:"name of the environment variable; must be a C_IDENTIFIER"`
	// Optional: defaults to "".
	Value string `json:"value,omitempty" description:"value of the environment variable; defaults to empty string"`
	// Optional; ObjectInfo represents a downward API request for a field of the object.
	//
	// Valid fields of this reference are:
	//
	// 1. APIVersion
	// 2. FieldPath
	//
	// No other fields may be specified.
	//
	// Currently ObjectInfo may be used to address the Name and Namespace
	// only.
	ObjectInfo *ObjectReference `json:"objectInfo,omitempty" description:"describes a field of the pod to use the value of; only APIVersion and FieldPath are accepted fields"`
	// Optional; HostInfo represents a downward API request for information about
	// the host.
	HostInfo *HostInfoSource `json:"hostInfo,omitempty" description:"describes information about the host to use"`
}
```

pmorie replied:
@bgrant0607 @smarterclayton I would be okay with having `ObjectInfo` become a field of `EnvVar` -- it would simplify the API greatly.

bgrant0607 replied:
If we think there may be multiple sources in the future, then I prefer the extra level of nesting. I was objecting to the specific combination of a pointer to an anonymous Go struct. Name the field Source.

pmorie replied:
@bgrant0607 Got it; I do expect that we will have additional sources in the future but I don't have another one I can offer at the moment besides ObjectInfo.  Under YAGNI, I vote to leave out VolumeSource until we have two.

bgrant0607 replied:
I'm fine with that.

bgrant0607 replied:
Perhaps we should call it fieldValue, though, rather than objectInfo, to strengthen the association with value.

thockin replied:
What do we want the JSON to look like?

```
{ 
    "name": "myvar",
    "source": { "objectInfo": { "fieldPath": ".metadata.name" } }
}
```

or

```
{ 
    "name": "myvar",
    "objectInfo": { "fieldPath": ".metadata.name" }
}
```

or 

```
{ 
    "name": "myvar",
    "value": { "objectInfo": {"fieldPath": ".metadata.name" }}
}
```

If I parse this correctly it's #2 that we are getting, yes?

thockin replied:
What about calling the sources something like "FromPod" or "ValueFromPod" ?

pmorie replied:
@thockin Either of those names is more descriptive -- unless we feel like we want open-ended info about objects, which I don't think we do.

bgrant0607 commented on pkg/api/v1beta3/types.go:
```diff
@@ -509,6 +509,33 @@ type EnvVar struct {
+	ObjectInfo *ObjectReference `json:"objectInfo,omitempty" description:"describes a field of the pod to use the value of; only APIVersion and FieldPath are accepted fields"`
+	// HostInfo represents a downward API request for information about
+	// the host.
+	HostInfo *HostInfoSource `json:"hostInfo,omitempty" description:"describes information about the host to use"`
```
HostInfo compromises isolation. Is there a concrete need for this? If not, I'd like to omit it.

pmorie replied:
@derekwaynecarr Did you have a concrete case for this?

pmorie replied:
@smarterclayton How about you -- got a concrete case for HostInfo?

derekwaynecarr replied:
@pmorie - I do not have an immediate need identified for HostInfo, I was more interested in ObjectInfo

bgrant0607 replied:
Let's omit HostInfo for now.

bgrant0607 commented on pkg/api/v1beta3/types.go:
```diff
@@ -509,6 +509,33 @@ type EnvVar struct {
+	//
+	// Currently ObjectInfo may be used to address the Name and Namespace
+	// only.
+	ObjectInfo *ObjectReference `json:"objectInfo,omitempty" description:"describes a field of the pod to use the value of; only APIVersion and FieldPath are accepted fields"`
```
I like the idea of using ObjectReference.

If fields from the ObjectReference are omitted, they should default to fields matching the pod.

Why is APIVersion necessary?

Right now, it looks like FieldPath is half-baked and only defined for one field:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubelet/container/ref.go#L47

We should make it match whatever syntax we settle on for kubectl #4348.

bgrant0607 replied:
Nevermind, APIVersion is needed in order to interpret FieldPath.

bgrant0607 replied:
But why shouldn't the APIVersion used for Pod itself be used?

smarterclayton replied:
If you reference another object it may not match the Kube API version (pods v1beta3 is different than Openshift build v1beta2)

> On Apr 14, 2015, at 5:06 PM, Brian Grant <notifications@github.com> wrote:
> 
> In pkg/api/v1beta3/types.go:
> 
> > +
> > +// EnvVarSource represents a source for the value of an EnvVar.
> > +// Only one of its members may be specified.
> > +type EnvVarSource struct {
> > +	// ObjectInfo represents a downward API request for a field of the object.
> > +	//
> > +	// Valid fields of this reference are:
> > +	//
> > +	// 1. APIVersion
> > +	// 2. FieldPath
> > +	//
> > +	// No other fields may be specified.
> > +	//
> > +	// Currently ObjectInfo may be used to address the Name and Namespace
> > +	// only.
> > +	ObjectInfo *ObjectReference `json:"objectInfo,omitempty" description:"describes a field of the pod to use the value of; only APIVersion and FieldPath are accepted fields"`
> But why shouldn't the APIVersion used for Pod itself be used?
> 
> —
> Reply to this email directly or view it on GitHub.
>

pmorie replied:
@bgrant0607 

> We should make it match whatever syntax we settle on for kubectl #4348.

Agree; are you okay with the way this PR handles name and namespace until that is worked out?

pmorie replied:
@smarterclayton @bgrant0607 

> If you reference another object it may not match the Kube API version (pods v1beta3 is different than Openshift build v1beta2)

Also, can't the API version change on the way out of the API server?

bgrant0607 replied:
Yes, like selfLink, we should really populate it based on the version requested, except in the case of cross-prefix references.

bgrant0607 replied:
If we don't allow Kind to be specified, I don't see why it makes sense to allow APIVersion to be specified.

pmorie replied:
@bgrant0607 

> If we don't allow Kind to be specified, I don't see why it makes sense to allow APIVersion to be specified.

My thinking was that the 'Kind' was limited to Pods.

pmorie commented:
Gonna go with this factoring in my next pass unless someone speaks soon:

@bgrant0607 Are you suggesting instead:

```go
// EnvVar represents an environment variable present in a Container.
type EnvVar struct {
	// Required: This must be a C_IDENTIFIER.
	Name string `json:"name" description:"name of the environment variable; must be a C_IDENTIFIER"`
	// Optional: defaults to "".
	Value string `json:"value,omitempty" description:"value of the environment variable; defaults to empty string"`
	// Optional; ObjectInfo represents a downward API request for a field of the object.
	//
	// Valid fields of this reference are:
	//
	// 1. APIVersion
	// 2. FieldPath
	//
	// No other fields may be specified.
	//
	// Currently ObjectInfo may be used to address the Name and Namespace
	// only.
	ObjectInfo *ObjectReference `json:"objectInfo,omitempty" description:"describes a field of the pod to use the value of; only APIVersion and FieldPath are accepted fields"`
}
```

Scratching `HostInfo`, we can add it back in later if there's a need.

thockin commented on pkg/api/types.go:
```diff
@@ -498,6 +498,33 @@ type EnvVar struct {
+	// 1. APIVersion
+	// 2. FieldPath
+	//
+	// No other fields may be specified.
```
This feels like an abuse of an existing type.  Why not just define your own type?

thockin commented on pkg/api/types.go:
```diff
@@ -498,6 +498,33 @@ type EnvVar struct {
+// EnvVarSource represents a source for the value of an EnvVar.
+// Only one of its members may be specified.
+type EnvVarSource struct {
+	// ObjectInfo represents a downward API request for a field of the object.
```
"of the object" - which object?  The Pod?  The Container?

thockin commented on pkg/api/v1beta1/conversion.go:
```diff
@@ -116,6 +116,26 @@ func init() {
+
+			return nil
+		},
+		func(in *newer.EnvVarSource, out *EnvVarSource, s conversion.Scope) error {
```
Why do you need this conversion?  Isn't it identical to the default?

pmorie replied:
@thockin This particular one exists because of the v1beta1 `Key` field of env var.

thockin commented on pkg/api/v1beta1/conversion.go:
```diff
@@ -116,6 +116,26 @@ func init() {
+			return nil
+		},
+		func(in *newer.HostInfoSource, out *HostInfoSource, s conversion.Scope) error {
+			out.HostName = in.HostName
```
Same question as above

pmorie replied:
@thockin I think you're right and these can go away.

thockin commented on pkg/api/v1beta1/conversion.go:
```diff
@@ -125,6 +145,26 @@ func init() {
+
+			return nil
+		},
+		func(in *EnvVarSource, out *newer.EnvVarSource, s conversion.Scope) error {
```
Same questions as above

thockin commented on pkg/api/types.go:
```diff
@@ -498,6 +498,33 @@ type EnvVar struct {
 	// Optional: defaults to "".
 	Value string `json:"value,omitempty"`
+	// Optional: specify a source for the value of the EnvVar.
+	*EnvVarSource `json:"source,omitempty"`
```
What if user specifies Value and one of the alt sources?  Who wins?  Document it.

pmorie replied:
@thockin In that case, it's a validation error; I will document.

thockin commented on pkg/api/validation/validation.go:
```diff
@@ -555,15 +555,80 @@ func validateEnv(vars []api.EnvVar) errs.ValidationErrorList {
+		} else if !util.IsCIdentifier(ev.Name) {
 			vErrs = append(vErrs, errs.NewFieldInvalid("name", ev.Name, cIdentifierErrorMsg))
 		}
+		vErrs = append(vErrs, validateEnvVarSources(ev)...)
```
Do you need to validate that one of Source or EnvVarSource are set, but not both?

thockin replied:
NM, see it below

thockin commented on pkg/api/validation/validation.go:
```diff
@@ -555,15 +555,80 @@ func validateEnv(vars []api.EnvVar) errs.ValidationErrorList {
+	}
+
+	if ev.Value != "" && numSources >= 1 {
+		allErrs = append(allErrs, errs.NewFieldInvalid("", "", "sources cannot be specified when value is set"))
```
error message could be more helpful.  "non-literal source may not be specified..." ?  Something a bit clearer.

thockin commented on pkg/api/validation/validation.go:
```diff
@@ -555,15 +555,80 @@ func validateEnv(vars []api.EnvVar) errs.ValidationErrorList {
+		allErrs = append(allErrs, validateHostInfo(ev.HostInfo)...)
+	}
+
+	if ev.Value != "" && numSources >= 1 {
```
s/>=1/> 0/ or even != 0

pmorie replied:
Yes, good catch

bgrant0607 commented on pkg/kubelet/kubelet.go:
```diff
@@ -759,6 +765,61 @@ func (kl *Kubelet) makeEnvironmentVariables(ns string, container *api.Container)
+	// an object using a fieldPath expression.  For now, support only the
+	// Name and Namespace fields.
+	switch internalFieldPath {
+	case "metadata.name":
```
I know we're currently using this syntax for other field specifications, but I'd prefer to go with something jsonpath-compatible. My reading of http://goessner.net/articles/JsonPath/ is that it would be: `$.metadata.name`. 

If we're going to break compatibility, we should do it sooner rather than later.

@lavalamp Any objections to changing field specifications to jsonpath syntax? Annoyingly, it's not the same as jq syntax, and there seem to be several other variants. However, all of the variants I've seen at least start paths with a `.`. Jsonpath and a few others start with `$.`. AFAICT, jsonpath is the most commonly mentioned.

pmorie replied:
This reminds me, there should really be a spec for what we come up with
that contains enough information to reimplement. The blog posts for
jsonpath aren't enough IMO. They're good blog posts but not a spec.
On Thu, Apr 16, 2015 at 12:05 AM Brian Grant <notifications@github.com>
wrote:

> In pkg/kubelet/kubelet.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6739#discussion_r28482739>
> :
>
> > +}
> > +
> > +func (kl *Kubelet) objectInfoRuntimeValue(objInfo *api.ObjectReference, pod *api.Pod) (string, error) {
> > +	// TODO: use objInfo Kind instead of "Pod"
> > +	internalFieldPath, _, err := api.Scheme.ConvertFieldLabel(objInfo.APIVersion, "Pod", objInfo.FieldPath, "")
> > +	if err != nil {
> > +		return "", err
> > +	}
> > +
> > +	runtimeVal := ""
> > +
> > +	// TODO: currently there is no generic utility for retrieving a field of
> > +	// an object using a fieldPath expression.  For now, support only the
> > +	// Name and Namespace fields.
> > +	switch internalFieldPath {
> > +	case "metadata.name":
>
> I know we're currently using this syntax for other field specifications,
> but I'd prefer to go with something jsonpath-compatible. My reading of
> http://goessner.net/articles/JsonPath/ is that it would be: $.
> metadata.name.
>
> If we're going to break compatibility, we should do it sooner rather than
> later.
>
> @lavalamp <https://github.com/lavalamp> Any objections to changing field
> specifications to jsonpath syntax? Annoyingly, it's not the same as jq
> syntax, and there seem to be several other variants. However, all of the
> variants I've seen at least start paths with a .. Jsonpath and a few
> others start with $.. AFAICT, jsonpath is the most commonly mentioned.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6739/files#r28482739>
> .
>

lavalamp replied:
@bgrant0607 I have no objection to moving to a more well defined field selector language.

thockin replied:
I am eager to have a more well defined selector language.  I am all for
specifications, but JSON Path doesn't seem to fit the bill as a spec (and
it has a bunch of stuff we may or may not want like wildcards, filters,
slices, scripts...)

Once we have this, I think we want to rapidly converge all sorts of
self-accessing things to it (e.g. the "metadata volume" PR).

Do we write our own spec that is a subset of jq and JSONPath?


On Wed, Apr 15, 2015 at 9:21 PM, Daniel Smith <notifications@github.com>
wrote:

> In pkg/kubelet/kubelet.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6739#discussion_r28483110>
> :
>
> > +}
> > +
> > +func (kl *Kubelet) objectInfoRuntimeValue(objInfo *api.ObjectReference, pod *api.Pod) (string, error) {
> > +	// TODO: use objInfo Kind instead of "Pod"
> > +	internalFieldPath, _, err := api.Scheme.ConvertFieldLabel(objInfo.APIVersion, "Pod", objInfo.FieldPath, "")
> > +	if err != nil {
> > +		return "", err
> > +	}
> > +
> > +	runtimeVal := ""
> > +
> > +	// TODO: currently there is no generic utility for retrieving a field of
> > +	// an object using a fieldPath expression.  For now, support only the
> > +	// Name and Namespace fields.
> > +	switch internalFieldPath {
> > +	case "metadata.name":
>
> @bgrant0607 <https://github.com/bgrant0607> I have no objection to moving
> to a more well defined field selector language.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6739/files#r28483110>
> .
>

smarterclayton replied:
I *think* we could define the subset as 

field names
array elements
map elements


foo.bar[1].baz
foo.bar[1]["baz"]

Where "baz" can be a map or field.  That's a strict subset of both JSONPath and jq to my knowledge.

----- Original Message -----
> > +}
> > +
> > +func (kl *Kubelet) objectInfoRuntimeValue(objInfo *api.ObjectReference,
> > pod *api.Pod) (string, error) {
> > +	// TODO: use objInfo Kind instead of "Pod"
> > +	internalFieldPath, _, err :=
> > api.Scheme.ConvertFieldLabel(objInfo.APIVersion, "Pod", objInfo.FieldPath,
> > "")
> > +	if err != nil {
> > +		return "", err
> > +	}
> > +
> > +	runtimeVal := ""
> > +
> > +	// TODO: currently there is no generic utility for retrieving a field of
> > +	// an object using a fieldPath expression.  For now, support only the
> > +	// Name and Namespace fields.
> > +	switch internalFieldPath {
> > +	case "metadata.name":
> 
> I am eager to have a more well defined selector language.  I am all for
> specifications, but JSON Path doesn't seem to fit the bill as a spec (and
> it has a bunch of stuff we may or may not want like wildcards, filters,
> slices, scripts...)
> 
> Once we have this, I think we want to rapidly converge all sorts of
> self-accessing things to it (e.g. the "metadata volume" PR).
> 
> Do we write our own spec that is a subset of jq and JSONPath?
> 
> 
> On Wed, Apr 15, 2015 at 9:21 PM, Daniel Smith <notifications@github.com>
> wrote:
> 
> > In pkg/kubelet/kubelet.go
> > <https://github.com/GoogleCloudPlatform/kubernetes/pull/6739#discussion_r28483110>
> > :
> >
> > > +}
> > > +
> > > +func (kl *Kubelet) objectInfoRuntimeValue(objInfo *api.ObjectReference,
> > > pod *api.Pod) (string, error) {
> > > +	// TODO: use objInfo Kind instead of "Pod"
> > > +	internalFieldPath, _, err :=
> > > api.Scheme.ConvertFieldLabel(objInfo.APIVersion, "Pod",
> > > objInfo.FieldPath, "")
> > > +	if err != nil {
> > > +		return "", err
> > > +	}
> > > +
> > > +	runtimeVal := ""
> > > +
> > > +	// TODO: currently there is no generic utility for retrieving a field
> > > of
> > > +	// an object using a fieldPath expression.  For now, support only the
> > > +	// Name and Namespace fields.
> > > +	switch internalFieldPath {
> > > +	case "metadata.name":
> >
> > @bgrant0607 <https://github.com/bgrant0607> I have no objection to moving
> > to a more well defined field selector language.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/GoogleCloudPlatform/kubernetes/pull/6739/files#r28483110>
> > .
> >
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/6739/files#r28487051

pmorie replied:
@thockin @smarterclayton @bgrant0607 @lavalamp 

A spec for a subset of jq/jsonpath is a good idea.  Tim is right that the 'spec' for jsonpath contains a ton of stuff we don't need and potentially may not want.

thockin replied:
Is it worthwhile to extract this to a pkg/fieldpath or pkg/api/fieldpath so that we can encapsulate later changes to the spec, and maybe offer validation hooks?

smarterclayton replied:
> On Apr 21, 2015, at 6:35 PM, Tim Hockin <notifications@github.com> wrote:
> 
> In pkg/kubelet/kubelet.go:
> 
> > +}
> > +
> > +func (kl *Kubelet) objectInfoRuntimeValue(objInfo *api.ObjectReference, pod *api.Pod) (string, error) {
> > +	// TODO: use objInfo Kind instead of "Pod"
> > +	internalFieldPath, _, err := api.Scheme.ConvertFieldLabel(objInfo.APIVersion, "Pod", objInfo.FieldPath, "")
> > +	if err != nil {
> > +		return "", err
> > +	}
> > +
> > +	runtimeVal := ""
> > +
> > +	// TODO: currently there is no generic utility for retrieving a field of
> > +	// an object using a fieldPath expression.  For now, support only the
> > +	// Name and Namespace fields.
> > +	switch internalFieldPath {
> > +	case "metadata.name":
> Is it worthwhile to extract this to a pkg/fieldpath or pkg/api/fieldpath so that we can encapsulate later changes to the spec, and maybe offer validation hooks?
> 
"meta" might be the most appropriate place because it abstracts the pattern for our objects without requiring the implementation.  Or even a separate file in pkg/api/ at minimum
> —
> Reply to this email directly or view it on GitHub.
>

pmorie replied:
I like encapsulating it in a package; will do.

thockin commented on pkg/api/validation/validation.go:
```diff
@@ -555,15 +555,80 @@ func validateEnv(vars []api.EnvVar) errs.ValidationErrorList {
+	if ref.APIVersion == "" {
+		allErrs = append(allErrs, errs.NewFieldRequired("apiVersion"))
+	}
+	if ref.Kind != "" {
```
wouldn't we do better to just ignore extra fields as if the struct did not have them?

thockin replied:
or to define a struct that doesn't have them

pmorie replied:
@thockin I think I agree with you.  How about:

```go
type FieldOfObject struct {
  APIVersion string
  FieldPath  string
}
```

pmorie commented:
Squashed, removed HostInfo, going to change `ObjectReference` to a custom type next.

pmorie commented:
@bgrant0607 @smarterclayton @thockin Feedback should be addressed; I removed the use of ObjectReference and added a narrower type, ObjectFieldSelector; added E2E test.  PTAL.

smarterclayton commented:
This looks pretty good now.

smarterclayton commented:
My original concerns have been addressed, it seems clear that you are selecting on a singleton, API versioning is clear.

bgrant0607 commented:
Please remove WIP from the title. Will look shortly.

pmorie commented:
Removed
On Tue, Apr 21, 2015 at 6:02 PM Brian Grant <notifications@github.com>
wrote:

> Please remove WIP from the title. Will look shortly.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6739#issuecomment-94956227>
> .
>

thockin commented on pkg/api/types.go:
```diff
@@ -508,6 +508,23 @@ type EnvVar struct {
 	// Optional: defaults to "".
 	Value string `json:"value,omitempty"`
+	// Optional: specify a source for the value of the EnvVar.
+	Source *EnvVarSource `json:"source,omitempty"`
```
Comment that Value and Source are mutually exclusive?

thockin commented on pkg/api/v1beta1/types.go:
```diff
@@ -397,6 +397,23 @@ type EnvVar struct {
 	// Optional: defaults to "".
 	Value string `json:"value,omitempty" description:"value of the environment variable; defaults to empty string"`
+	// Optional: specify a source for the value of the EnvVar.
+	Source *EnvVarSource `json:"source,omitempty" description:"source for the environment variable's value; cannot be used if value is not empty"`
```
I feel like this could be named more obviously, still.

ValueFrom?  Eval?

I'm just thinking about how the JSON reads - what's the story that it tells?  "Name 'foo', value from pod field v1beta1.ID" is a pretty good story, better than ""Name 'foo', source pod field v1beta1.ID".  and it ValueFrom makes the mutual exclusion more obvious.

pmorie replied:
Do want `ValueFrom`, do not want `Eval`, I will change it

thockin commented on pkg/api/v1beta1/types.go:
```diff
@@ -397,6 +397,23 @@ type EnvVar struct {
+// ObjectFieldSelector selects an APIVersioned field of an object.
+type ObjectFieldSelector struct {
+	// The API version the FieldPath is written in terms of.
+	APIVersion string `json:"apiVersion,omitempty" description="The API version that FieldPath is written in terms of"`
```
This is fine for now - it's not your fault/problem, but I want to be on record that asking the user to spec an API version inside an object that is already API-versioned is asinine.  We should endeavor to do better than this in API v2.

smarterclayton replied:
> On Apr 21, 2015, at 6:28 PM, Tim Hockin <notifications@github.com> wrote:
> 
> In pkg/api/v1beta1/types.go:
> 
> > @@ -397,6 +397,23 @@ type EnvVar struct {
> >  	Key  string `json:"key,omitempty" description:"name of the environment variable; must be a C_IDENTIFIER; deprecated - use name instead"`
> >  	// Optional: defaults to "".
> >  	Value string `json:"value,omitempty" description:"value of the environment variable; defaults to empty string"`
> > +	// Optional: specify a source for the value of the EnvVar.
> > +	Source *EnvVarSource `json:"source,omitempty" description:"source for the environment variable's value; cannot be used if value is not empty"`
> > +}
> > +
> > +// EnvVarSource represents a source for the value of an EnvVar.
> > +// Only one of its members may be specified.
> > +type EnvVarSource struct {
> > +	// Selects a field of the pod; only name and namespace are supported.
> > +	PodField *ObjectFieldSelector `json:"podField,omitempty" description:"selectes a field of the pod; only name and namespace are supported"`
> > +}
> > +
> > +// ObjectFieldSelector selects an APIVersioned field of an object.
> > +type ObjectFieldSelector struct {
> > +	// The API version the FieldPath is written in terms of.
> > +	APIVersion string `json:"apiVersion,omitempty" description="The API version that FieldPath is written in terms of"`
> This is fine for now - it's not your fault/problem, but I want to be on record that asking the user to spec an API version inside an object that is already API-versioned is asinine. We should endeavor to do better than this in API v2.
> 
All we have to do is default it in the conversion.  We just don't have a cheap way of doing that.
> —
> Reply to this email directly or view it on GitHub.
>

bgrant0607 replied:
Same problem as ObjectReference, validation errors, field selectors, and other field references.

Sorry I missed the intervening discussion (too much going on), but podField and ObjectFieldSelector seem inelegant. ObjectReference was sufficient and more extensible. I understand that it seems overly general right now, but if we wanted more specific, I'd rather scrap both EnvVarSource and ObjectFieldSelector and just put ValueFromField in EnvVar.

thockin replied:
regarding ObjectRef vs ObjectFieldSelector, here's my thinking.

It is a strict subset of ObejctRef.  We should not reuse struct just because it exists.  If it grows to be identical to ObjectRef, we can swap in that struct at will, and it will be compatible.  We would not add fields to a struct just because we might need them.  I feel we should be doing this MORE - use purpose-built structs that fit exactly what we need.  The struct name is not really even part of the API.

That's the last I'll say, if Brian wants to use Objectref, just do that

pmorie replied:
@bgrant0607 is `ValueFromField` a string?
On Wed, Apr 22, 2015 at 12:30 AM Brian Grant <notifications@github.com>
wrote:

> In pkg/api/v1beta1/types.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6739#discussion_r28843747>
> :
>
> > @@ -397,6 +397,23 @@ type EnvVar struct {
> >  	Key  string `json:"key,omitempty" description:"name of the environment variable; must be a C_IDENTIFIER; deprecated - use name instead"`
> >  	// Optional: defaults to "".
> >  	Value string `json:"value,omitempty" description:"value of the environment variable; defaults to empty string"`
> > +	// Optional: specify a source for the value of the EnvVar.
> > +	Source *EnvVarSource `json:"source,omitempty" description:"source for the environment variable's value; cannot be used if value is not empty"`
> > +}
> > +
> > +// EnvVarSource represents a source for the value of an EnvVar.
> > +// Only one of its members may be specified.
> > +type EnvVarSource struct {
> > +	// Selects a field of the pod; only name and namespace are supported.
> > +	PodField *ObjectFieldSelector `json:"podField,omitempty" description:"selectes a field of the pod; only name and namespace are supported"`
> > +}
> > +
> > +// ObjectFieldSelector selects an APIVersioned field of an object.
> > +type ObjectFieldSelector struct {
> > +	// The API version the FieldPath is written in terms of.
> > +	APIVersion string `json:"apiVersion,omitempty" description="The API version that FieldPath is written in terms of"`
>
> Same problem as ObjectReference, validation errors, field selectors, and
> other field references.
>
> Sorry I missed the intervening discussion (too much going on), but
> podField and ObjectFieldSelector seem inelegant. ObjectReference was
> sufficient and more extensible. I understand that it seems overly general
> right now, but if we wanted more specific, I'd rather scrap both
> EnvVarSource and ObjectFieldSelector and just put ValueFromField in EnvVar.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6739/files#r28843747>
> .
>

bgrant0607 replied:
If you want to use ObjectFieldSelector, document that it must be a strict subset of ObjectReference.

My bigger beef is with the field name: podField. I'm 99% certain that we're going to need to be able to substitute fields of other objects. In the case that we support that in the API, podField is awkward. In the case that we push cross-object substitution to the client, podField seems both over- and under-specified (of course it must be a pod, but it can't be another pod).

I'll try to provide a more constructive suggestion later tonight.

bgrant0607 replied:
What's the argument against just "field"?

bgrant0607 replied:
description nits: Delete leading "The" and refer to the json field names, not the Go struct names (so, fieldPath rather than FieldPath).

bgrant0607 replied:
Ping (here and other API versions)

thockin commented on pkg/api/validation/validation.go:
```diff
@@ -555,15 +555,50 @@ func validateEnv(vars []api.EnvVar) errs.ValidationErrorList {
+	return allErrs
+}
+
+func validateObjectFieldSelector(fs *api.ObjectFieldSelector) errs.ValidationErrorList {
```
At what point do we validate these are legitimate values?

pmorie replied:
@thockin At this point, that happens in the kubelet.  I don't think that's great.  Let me think on this a bit more.

bgrant0607 replied:
Seems totally feasible to validate them in apiserver even if expanded by Kubelet.

smarterclayton replied:
> On Apr 22, 2015, at 9:30 PM, Brian Grant <notifications@github.com> wrote:
> 
> In pkg/api/validation/validation.go:
> 
> > +	numSources := 0
> > +
> > +	switch {
> > +	case ev.Source.PodField != nil:
> > +		numSources++
> > +		allErrs = append(allErrs, validateObjectFieldSelector(ev.Source.PodField).Prefix("podField")...)
> > +	}
> > +
> > +	if ev.Value != "" && numSources != 0 {
> > +		allErrs = append(allErrs, errs.NewFieldInvalid("", "", "sources cannot be specified when value is not empty"))
> > +	}
> > +
> > +	return allErrs
> > +}
> > +
> > +func validateObjectFieldSelector(fs *api.ObjectFieldSelector) errs.ValidationErrorList {
> Seems totally feasible to validate them in apiserver even if expanded by Kubelet.
> 
So if an network plugin was running on the nodes and could offer downward API to pods (not saying I have a concrete scenario in mind), the apiserver would need to know about the full shape of the downward API object the network plugin exposes?  Just trying to clarify what we mean by validation.
> —
> Reply to this email directly or view it on GitHub.
>

bgrant0607 replied:
The user would prefer to get errors as early as possible. The plugin wouldn't have to be compiled into the apiserver or kubectl in order to enable that; we could use swagger.

thockin commented on pkg/api/v1beta1/types.go:
```diff
@@ -397,6 +397,23 @@ type EnvVar struct {
+	// The API version the FieldPath is written in terms of.
+	APIVersion string `json:"apiVersion,omitempty" description="The API version that FieldPath is written in terms of"`
+	// The path of the field to select in the specified API version
+	FieldPath string `json:"fieldPath,omitempty" description="The path of the field to select in the specified API version"`
```
Need to document the format of this somewhere

bgrant0607 replied:
+1 to documentation. For starters, an example in the description `(e.g., "metadata.name")` would be useful.

Delete leading "The".

FieldPath isn't really optional.

bgrant0607 replied:
Ping (here and other API versions)

pmorie replied:
Gah, `description=`

pmorie commented:
@thockin @bgrant0607 Spent some time thinking about this, and here's what I suggest:

1.  Make `ValueFromPodField` a field of `EnvVar`
2.  The semantics of `ValueFromPodField` are that it is in terms of the APIVersion of the pod
    of the pod:
    1.  It is translated during the conversion between API versions
    2.  It is validated at the API server as part of the PodSpec
3.  Validation of the fieldPath in PodSpec validation
4.  Add interfaces for a facility to get the value at a field path from an object and an initial impl that supports metadata.name and metadata.namespace for PodSpecs.

```go
type EnvVar struct {
  Name              string
  Value             string
  ValueFromPodField string
}
```

API looks like:

```yaml
apiVersion: v1beta3
kind: Pod
metadata:
  name: www
spec:
  containers:
  - name: example
    image: example
    env:
    - name: GIT_REPO
      valueFromPodField: metadata.name
```

This gets us to:

1.  Ok for-now API
2.  Versioning of the expression is solved
3.  FieldPath encapsulation can be reused even if we refactor the API
4.  FieldPath expression language can be implemented in future to a spec we write.

smarterclayton commented:
I don't like value from PodField - I don't think adding that field is any better than your previous.  I'm ok with ObjectReference or ObjectFieldSelector - whatever will have minimal impact later on.  Since Selector is a strict subset we can always change it.  I don't think forcing a single field makes things better.

pmorie commented:
@smarterclayton If we handle the API versioning correctly, there's no need for there to be anything but a string.  I can do:

```go
type ObjectFieldSelector struct {
  FieldPath string
}
```

If that will be better.

smarterclayton commented:
You have to be able to specify an APIVersion - it might defaulted to empty internally, but as soon as we have other object types we have to add APIVersion in, and I'm much more concerned about that person not realizing that breaks pods.

----- Original Message -----
> @smarterclayton If we handle the API versioning correctly, there's no need
> for there to be anything but a string.  I can do:
> 
> ```go
> type ObjectFieldSelector struct {
>   FieldPath string
> }
> ```
> 
> If that will be better.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/6739#issuecomment-95323030

pmorie commented:
@smarterclayton Why do we have to add APIVersion once you have other types?

smarterclayton commented:
Because the pod api version has nothing to do with the other objects version.  You can only convert APIversion on your own type (really).

----- Original Message -----
> @smarterclayton Why do we have to add APIVersion once you have other types?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/6739#issuecomment-95327965

pmorie commented:
@smarterclayton Got it

pmorie commented:
Alright, with that in mind... I propose to:

1.  Keep the type the way it is now
2.  Introduce the fieldpath abstractions and validation / transformation in the API server
3.  Additional detail: if the API version of an ObjectFieldSelector is unspecified, default it to the APIVersion being converted from.

That way, we can have the APIVersion for future use, and make the API sensible.  If you don't specify an API version, it will be defaulted from the one being converted from on the way into the internal API.

smarterclayton commented:
Yes 3 makes total sense.

pmorie commented:
OK

thockin commented:
@pmorie 

> The semantics ... are that it is in terms of the APIVersion of the pod of the pod

This is what I want but I don't think we have time to actually implement that right now.  I'm not sure I buy that cross-object field-paths need an API version.  That sets a trap for users who will be required to eventually update their object just to indicate a different API version, even if no semantic change.  That stinks, to me.

As long as a field exists in the API, a long-lived object which references it should not have to change.  To me, that says that these field-path references would be converted from versioned to internal and back, just like objects are.  At some point stored objects need to be updated to newere stored forms, and the field-path should just keep working.  When the target field is finally deprecated, it should be an error to continue to use those fields.

No need to "refresh" objects with cross-object refs.

This is NOT a 1.0 goal.

smarterclayton commented:
> On Apr 22, 2015, at 7:55 PM, Tim Hockin <notifications@github.com> wrote:
> 
> @pmorie
> 
> The semantics ... are that it is in terms of the APIVersion of the pod of the pod
> 
> This is what I want but I don't think we have time to actually implement that right now. I'm not sure I buy that cross-object field-paths need an API version. That sets a trap for users who will be required to eventually update their object just to indicate a different API version, even if no semantic change. That stinks, to me.
> 
> As long as a field exists in the API, a long-lived object which references it should not have to change. To me, that says that these field-path references would be converted from versioned to internal and back, just like objects are. At some point stored objects need to be updated to newere stored forms, and the field-path should just keep working. When the target field is finally deprecated, it should be an error to continue to use those fields.
> 
> No need to "refresh" objects with cross-object refs.
> 
Are you assuming Kube apiserver knows about the object the field path is in?  Because I have been assuming it does not, and that only the kubelet and client ultimately may know about the object referenced here.
> This is NOT a 1.0 goal.
> 
> —
> Reply to this email directly or view it on GitHub.
>

thockin commented:
I am assuming that "whoever stores the durable form of an object" can do
the hub-and-spokes version conversion, and that includes field-paths (ergo,
they aren't strings but a formal type).

On Wed, Apr 22, 2015 at 5:21 PM, Clayton Coleman <notifications@github.com>
wrote:

>
>
> > On Apr 22, 2015, at 7:55 PM, Tim Hockin <notifications@github.com>
> wrote:
> >
> > @pmorie
> >
> > The semantics ... are that it is in terms of the APIVersion of the pod
> of the pod
> >
> > This is what I want but I don't think we have time to actually implement
> that right now. I'm not sure I buy that cross-object field-paths need an
> API version. That sets a trap for users who will be required to eventually
> update their object just to indicate a different API version, even if no
> semantic change. That stinks, to me.
> >
> > As long as a field exists in the API, a long-lived object which
> references it should not have to change. To me, that says that these
> field-path references would be converted from versioned to internal and
> back, just like objects are. At some point stored objects need to be
> updated to newere stored forms, and the field-path should just keep
> working. When the target field is finally deprecated, it should be an error
> to continue to use those fields.
> >
> > No need to "refresh" objects with cross-object refs.
> >
> Are you assuming Kube apiserver knows about the object the field path is
> in? Because I have been assuming it does not, and that only the kubelet and
> client ultimately may know about the object referenced here.
> > This is NOT a 1.0 goal.
> >
> > —
> > Reply to this email directly or view it on GitHub.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6739#issuecomment-95373789>
> .
>

smarterclayton commented:
That requires pods to know about all possible objects in the system though (if by stores you mean "apiserver"), which I thought we think is an anti-goal.  I think of that value as a passthrough intended for the kubelet, like volumes.

How does someone extend the downward API without changes to the apiserver?

> On Apr 22, 2015, at 8:44 PM, Tim Hockin <notifications@github.com> wrote:
> 
> I am assuming that "whoever stores the durable form of an object" can do
> the hub-and-spokes version conversion, and that includes field-paths (ergo,
> they aren't strings but a formal type).
> 
> On Wed, Apr 22, 2015 at 5:21 PM, Clayton Coleman <notifications@github.com>
> wrote:
> 
> >
> >
> > > On Apr 22, 2015, at 7:55 PM, Tim Hockin <notifications@github.com>
> > wrote:
> > >
> > > @pmorie
> > >
> > > The semantics ... are that it is in terms of the APIVersion of the pod
> > of the pod
> > >
> > > This is what I want but I don't think we have time to actually implement
> > that right now. I'm not sure I buy that cross-object field-paths need an
> > API version. That sets a trap for users who will be required to eventually
> > update their object just to indicate a different API version, even if no
> > semantic change. That stinks, to me.
> > >
> > > As long as a field exists in the API, a long-lived object which
> > references it should not have to change. To me, that says that these
> > field-path references would be converted from versioned to internal and
> > back, just like objects are. At some point stored objects need to be
> > updated to newere stored forms, and the field-path should just keep
> > working. When the target field is finally deprecated, it should be an error
> > to continue to use those fields.
> > >
> > > No need to "refresh" objects with cross-object refs.
> > >
> > Are you assuming Kube apiserver knows about the object the field path is
> > in? Because I have been assuming it does not, and that only the kubelet and
> > client ultimately may know about the object referenced here.
> > > This is NOT a 1.0 goal.
> > >
> > > —
> > > Reply to this email directly or view it on GitHub.
> > >
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/GoogleCloudPlatform/kubernetes/pull/6739#issuecomment-95373789>
> > .
> >
> —
> Reply to this email directly or view it on GitHub.
>

bgrant0607 commented on pkg/api/v1beta3/conversion.go:
```diff
@@ -28,6 +28,7 @@ func init() {
 		func(label, value string) (string, string, error) {
 			switch label {
 			case "metadata.name",
+				"metadata.namespace",
```
I'm resigned to the fact that starting with this syntax is the most expedient thing to do. We can reevaluate later.

In the long run, we should strive to make all field references the same:
* downward API substitution
* client-side cross-object field references
* kubectl -o template
* field selectors
* ObjectReference
* field projection (only return a subset of object fields)

bgrant0607 replied:
Forget I said anything. :-) I'd like to get this in.

bgrant0607 commented on pkg/api/v1beta1/types.go:
```diff
@@ -397,6 +397,23 @@ type EnvVar struct {
+// Only one of its members may be specified.
+type EnvVarSource struct {
+	// Selects a field of the pod; only name and namespace are supported.
+	PodField *ObjectFieldSelector `json:"podField,omitempty" description:"selectes a field of the pod; only name and namespace are supported"`
```
Typo: selectes

bgrant0607 replied:
Let's call this field `fieldPath`.

thockin commented:
@smarterclayton This is a moot point for now, and it is hard, so let's not spend too much time on it now.

> That requires pods to know about all possible objects in the system though (if by stores you mean "apiserver"), which I thought we think is an anti-goal.  I think of that value as a passthrough intended for the kubelet, like volumes.

Yeah, so truly opaque API blobs that get passed all the way down to Kubelet is gonna suck for things like field-references.  I really hate the idea of laying time bombs that have to be refreshed by the user for zero purpose other than we're too dumb to figure out what they meant.  Maybe you just can't do "master-evaluated" fields in opaque blobs.  Not that we support opaque blobs yet.

pmorie commented:
We can sidestep this point entirely for now if we go with a super simple API like so:

```go
type EnvVarSource struct {
  PodName      bool
  PodNamespace bool
}
```

It's expressive enough:

```yaml
apiVersion: v1beta3
kind: Pod
metadata:
  name: www
spec:
  containers:
  - name: example
    image: example
    env:
    - name: POD_NAME
      source:
        podName: true
    - name: POD_NAMESPACE
      source:
        podNamespace: true
```

Just putting that out there.  I don't think anyone's especially happy with introducing
the syntax for the fields.  I don't especially like it personally.

Any thoughts? @thockin @bgrant0607 @smarterclayton

thockin commented:
I'm not above making up a different way to do this, but I thought we were
close to agreement on syntax for fields refs?

On Wed, Apr 22, 2015 at 10:00 PM, Paul Morie <notifications@github.com>
wrote:

> We can sidestep this point entirely for now if we go with a super simple
> API like so:
>
> type EnvVarSource struct {
>   PodName      bool
>   PodNamespace bool
> }
>
> It's expressive enough:
>
> apiVersion: v1beta3kind: Podmetadata:
>   name: wwwspec:
>   containers:
>   - name: example
>     image: example
>     env:
>     - name: POD_NAME
>       source:
>         podName: true
>     - name: POD_NAMESPACE
>       source:
>         podNamespace: true
>
> Just putting that out there. I don't think anyone's especially happy with
> introducing
> the syntax for the fields. I don't especially like it personally.
>
> Any thoughts? @thockin <https://github.com/thockin> @bgrant0607
> <https://github.com/bgrant0607> @smarterclayton
> <https://github.com/smarterclayton>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6739#issuecomment-95432762>
> .
>

pmorie commented:
@thockin Your arguments convinced me that we were getting ourselves in
deeper on this feature than it would be worth. At the end of the day it
seems like a lot of complexity for an api that covers two fields at the
moment. Since the expression syntax isn't a 1.0 goal I thought I would
throw an alternative out. I think ultimately it is best to use a field
path, but as @bgrant0607 pointed out there are a ton of cross-cuts to be
considered with all the use-cases.

So, with all that said, if we want to go with a field path impl at this
moment, let's talk through how to handle the eventual migration of these
blobs.

The semantics we've discussed have been:

1. APIVersion is stored in the selector and defaulted on creation to be the
APIVersion of the enclosing object
2. The field path is converted as part of the conversion into the internal
API and validated at the API server
3. The field path is converted to the selector's APIVersion when converting
out if the internal api

So, a serialized pod spec in the store will have an internal API field path
and the original APIVersion to reify the field path. If you have a
migration operation that migrates that object's internal APIVersion, that
operation can also transform the field path between the old internal api
and the new internal api. It's entirely possible that my mental model is
broken but the use-case of migrating the field paths doesn't seem
especially difficult to me.

Also, if we go with field path, we will want a high quality skeleton and
spec for what we currently support. The spec can be really simple;
technically we don't even need to spec out array access or maps for this
case at the moment. There can be a note in the spec that only the metadata
name and namespace are supported. We can implement features as we need them
for other use cases and update the spec as we go.

I'm happy to do the work necessary for a solid field path skeleton now if
that is what the group wants. To lay it all out, the ingredients would be:

1. Validation
2. Skeletal spec for access to string fields, no array or map syntax and
accompanying skeletal design doc
3. Basic implementation that supports metadata name and namespace

Show of hands, who would prefer to do a simple field path? Who would prefer
ultra-basic API?

On Thu, Apr 23, 2015 at 1:13 AM Tim Hockin <notifications@github.com> wrote:

> I'm not above making up a different way to do this, but I thought we were
> close to agreement on syntax for fields refs?
>
> On Wed, Apr 22, 2015 at 10:00 PM, Paul Morie <notifications@github.com>
> wrote:
>
> > We can sidestep this point entirely for now if we go with a super simple
> > API like so:
> >
> > type EnvVarSource struct {
> > PodName bool
> > PodNamespace bool
> > }
> >
> > It's expressive enough:
> >
> > apiVersion: v1beta3kind: Podmetadata:
> > name: wwwspec:
> > containers:
> > - name: example
> > image: example
> > env:
> > - name: POD_NAME
> > source:
> > podName: true
> > - name: POD_NAMESPACE
> > source:
> > podNamespace: true
> >
> > Just putting that out there. I don't think anyone's especially happy with
> > introducing
> > the syntax for the fields. I don't especially like it personally.
> >
> > Any thoughts? @thockin <https://github.com/thockin> @bgrant0607
> > <https://github.com/bgrant0607> @smarterclayton
> > <https://github.com/smarterclayton>
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/GoogleCloudPlatform/kubernetes/pull/6739#issuecomment-95432762
> >
> > .
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6739#issuecomment-95435002>
> .
>

smarterclayton commented:
I think the agreement was:

```
  env:
  - name:
    source:
      kind: Pod
      fieldPath: ....
```

with 

```
type EnvVarSource struct {
  *ObjectFieldSource
}
type ObjectFieldSource struct {
  Kind
  FieldPath
}
```

@thockin that match your expectation?

bgrant0607 commented:
@smarterclayton, @thockin, and I discussed this just now. We prefer the simple field path. Let's call it `fieldPath`, and use `ObjectFieldSelector`, which must be a strict subset of `ObjectReference`. Also, let's rename `source` to `valueFrom`.

thockin commented:
`valueFrom: { fieldPath: { apiVersion: v1beta1, fieldPath: metadata.name} }`

SGTM

On Thu, Apr 23, 2015 at 11:18 AM, Brian Grant <notifications@github.com>
wrote:

> @smarterclayton <https://github.com/smarterclayton>, @thockin
> <https://github.com/thockin>, and I discussed this just now. We prefer
> the simple field path. Let's call it fieldPath, and use
> ObjectFieldSelector, which must be a strict subset of ObjectReference.
> Also, let's rename source to valueFrom.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6739#issuecomment-95675877>
> .
>

pmorie committed:
Add downward API for environment vars

pmorie commented:
Changed the field name, next is factoring the fieldpath stuff out correctly.

pmorie commented:
@bgrant0607 Should be ready for final review now, PTAL

pmorie commented:
@bgrant0607 New e2e runs locally for me

pmorie commented:
Ok, let's try this again.  A change went in to master that didn't create a merge conflict but broke the serialization test locally.

bgrant0607 commented on pkg/api/v1beta1/defaults_test.go:
```diff
@@ -294,3 +294,29 @@ func TestSetDefaultMinionExternalID(t *testing.T) {
+
+	apiVersion := s2.Containers[0].Env[0].ValueFrom.FieldPath.APIVersion
+	if apiVersion != "v1beta1" {
+		t.Errorf("Expected default APIVersion v1beta2, got: %v", apiVersion)
```
Expected v1beta1

pmorie replied:
How embarrassing @bgrant0607

bgrant0607 commented on pkg/api/v1beta3/defaults_test.go:
```diff
@@ -302,3 +302,31 @@ func TestSetDefaultNodeExternalID(t *testing.T) {
+			{
+				Env: []current.EnvVar{
+					{
+
```
Unnecessary blank line.

bgrant0607 commented:
Look pretty good! Just a few minor comments.

pmorie commented:
@bgrant0607 Fixed

bgrant0607 commented:
2 comments overlooked, I think.

Looking forward to merging this on Monday. Thanks for pushing this forward.

pmorie commented:
@bgrant0607 All comments should be fixed now

smarterclayton commented:
So now that this awesome feature is almost done... how do I choose my own names for environment variables for services?

pmorie commented:
@smarterclayton We can implement a new env var source for that info

smarterclayton commented:
Is it a different source?  Seems like it's the spec.portalIP and spec.port[0] field of a service with a given name and namespace.

pmorie commented:
@smarterclayton 

> with a given name and namespace.

Wouldn't it just be the name, since we don't support cross-ns services?

smarterclayton commented:
It could be eventually.  But yes, for now, just name.

----- Original Message -----
> @smarterclayton
> 
> > with a given name and namespace.
> 
> Wouldn't it just be the name, since we don't support cross-ns services?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/6739#issuecomment-96723979

pmorie commented:
@smarterclayton That sounds to me like:

```go
type EnvVarSource struct {
  FieldPath        *ObjectFieldSelector
  ServiceFieldPath *NamedObjectFieldSelector
}

type NamedObjectFieldSelector {
	ObjectFieldSelector

	Name string
}
```

pmorie commented:
Kicked shippable

smarterclayton commented:
And then we say... why do we have ServiceFieldPath and FieldPath together?

----- Original Message -----
> @smarterclayton That sounds to me like:
> 
> ```go
> type EnvVarSource struct {
>   FieldPath        *ObjectFieldSelector
>   ServiceFieldPath *NamedObjectFieldSelector
> }
> 
> type NamedObjectFieldSelector {
> 	ObjectFieldSelector
> 
> 	Name string
> }
> ```
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/6739#issuecomment-96728866

pmorie commented:
@smarterclayton this would definitely lurch us in the direction of `ObjectReference`

smarterclayton commented:
Yes.  I'm not proposing any changes for this pull (since the types are the same) but I realized we got really focused on pod name/namespace.  The follow-on will need to deal with the implications.

----- Original Message -----
> @smarterclayton this would definitely lurch us in the direction of
> `ObjectReference`
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/6739#issuecomment-96729225

pmorie commented:
@smarterclayton Agree, I will make a new issue for service downward API

thockin commented:
Services wanders into a whole different discussion - the pre-declaring
service-deps topic.  I don't yet have an opinion as to whether we should
have an env var source way to do this, but I do think it is more
complicated that you have listed here.

Service have multiple ports - do I have to call out each one I want?  How
do I express named ports, rather than indexed? etc.

On Mon, Apr 27, 2015 at 9:20 AM, Paul Morie <notifications@github.com>
wrote:

> @smarterclayton <https://github.com/smarterclayton> Agree, I will make a
> new issue for service downward API
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6739#issuecomment-96729690>
> .
>

pmorie commented:
@thockin I am looking at the downward API for service envs as orthogonal to pre-declaring service dependencies.  I think dependencies will take pre-start hooks and should be separate in the API from how the vars are constructed.

smarterclayton commented:
Exactly.  But exposing services via this mechanism is strictly better in my opinion than making up a bunch of env vars (making up one env var I can get behind, just not 100 different variants).

----- Original Message -----
> Services wanders into a whole different discussion - the pre-declaring
> service-deps topic.  I don't yet have an opinion as to whether we should
> have an env var source way to do this, but I do think it is more
> complicated that you have listed here.
> 
> Service have multiple ports - do I have to call out each one I want?  How
> do I express named ports, rather than indexed? etc.
> 
> On Mon, Apr 27, 2015 at 9:20 AM, Paul Morie <notifications@github.com>
> wrote:
> 
> > @smarterclayton <https://github.com/smarterclayton> Agree, I will make a
> > new issue for service downward API
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/GoogleCloudPlatform/kubernetes/pull/6739#issuecomment-96729690>
> > .
> >
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/6739#issuecomment-96731151

pmorie commented:
Ugh. Rebasing

pmorie commented:
Rebased

pmorie commented:
Local e2e run looks good.

pmorie commented:
@bgrant0607 Does this need anything else?

bgrant0607 commented on pkg/api/v1beta1/defaults.go:
```diff
@@ -166,6 +166,11 @@ func init() {
 				obj.ExternalID = obj.ID
 			}
 		},
+		func(obj *ObjectFieldSelector) {
```
Did you confirm that this isn't called when the pointer is nil?

bgrant0607 commented:
The ObjectFieldSelector field descriptions didn't get changed as I asked, but my review latency is very high right now, so I'm not going to block on that.

LGTM.

pmorie commented:
@bgrant0607 Dang, I thought I had absolutely everything.  I will fix them in a follow-up.
